rest-server/test/userManager.js

```diff
@@ -0,0 +1,388 @@
+
+
+//
+// Get a valid token that expires in 60 seconds.
+//
+
+const validToken = global.jwt.sign({ username: 'new_user', admin: true }, process.env.JWT_SECRET, { expiresIn: 60 });
+const invalidToken = '';
+
+describe('Add new user: put /api/v1/user', () => {
+  beforeEach(() => {
+
+    nock(etcdHosts)
+      .get('/v2/keys/users/new_user')
+      .reply(200, {
+        'errorCode': 100,
+        'message': 'Key not found',
+        'cause': '/test',
+        'index': 1
+      });
+
+    nock(etcdHosts)
+      .put('/v2/keys/users/new_user?dir=true')
+      .reply(200, {
+        'action': 'set',
+        'node': {
+          'key': '/users/new_user',
+          'dir': true,
+          'modifiedIndex': 2,
+          'createdIndex': 2
+        }
+      });
+
+    nock(etcdHosts)
+      .put('/v2/keys/users/new_user/passwd', { 'value': '8507b5d862306d5bdad95b3d611b905ecdd047b0165ca3905db0d861e76bce8f3a8046e64379e81f54865f7c41b47e57cec887e5912062211bc9010afea8ab12' })
+      .reply(200, {
+        'action': 'set',
+        'node': {
+          'key': '/users/new_user/password',
+          'value': '8507b5d862306d5bdad95b3d611b905ecdd047b0165ca3905db0d861e76bce8f3a8046e64379e81f54865f7c41b47e57cec887e5912062211bc9010afea8ab12',
+          'modifiedIndex': 3,
+          'createdIndex': 3
+        }
+      });
+
+    nock(etcdHosts)
+      .put('/v2/keys/users/new_user/admin', { 'value': 'true' })
+      .reply(200, {
+        'action': 'set',
+        'node': {
+          'key': '/users/new_user/admin',
+          'value': 'true',
+          'modifiedIndex': 4,
+          'createdIndex': 4
+        }
+      });
+
+  });
+
+  //
+  // Positive cases
+  //
+
+  it('Case 1 (Positive): Add a admin user with modify is false.', (done) => {
+    global.chai.request(global.server)
+      .put('/api/v1/user')
+      .set('Authorization', 'Bearer ' + validToken)
+      .send(JSON.parse(global.mustache.render(newUserTemplate, { 'username': 'new_user', 'password': '123456', 'admin': true, 'modify': false })))
+      .end((err, res) => {
+        global.chai.expect(res, 'status code').to.have.status(201);
+        global.chai.expect(res, 'response format').be.json;
+        global.chai.expect(res.body.message, 'response message').equal('update successfully');
+        done();
+      });
+  });
+
+  //
+  // Negative cases
+  //
+
+  it('Case 2 (Negative): Add a new admin user failed because modify value is true.', (done) => {
+    global.chai.request(global.server)
+      .put('/api/v1/user')
+      .set('Authorization', 'Bearer ' + validToken)
+      .send(JSON.parse(global.mustache.render(newUserTemplate, { 'username': 'new_user', 'password': '123456', 'admin': true, 'modify': true })))
+      .end((err, res) => {
+        global.chai.expect(res, 'status code').to.have.status(500);
+        global.chai.expect(res, 'response format').be.json;
+        global.chai.expect(res.body.message, 'response message').equal('update failed');
+        done();
+      });
```

Comment:
Add another negative case: what if the user that calls the api is not an admin?

Reply:
Add negative case to add, update and delete user with non-admin token.